Fix `ToJSON` instance of `BoundedRatio` to avoid precision loss

Lucsanszky opened the pull request:
# Description
Resolves #4026 
<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

Lucsanszky committed:
Fix `ToJSON` precision loss of `BoundedRatio`

* Avoid losing precision when converting to JSON from a `BoundedRatio`
by falling back to the `Rational` JSON instance when there are too many
repetends

* Update the `FromJSON` instance of `BoundedRatio` to conform to the above

* Move `toRationalJSON` to `BaseTypes`

* Remove tests that are now obsolete due to the above changes

* Factor out `fromRationalEither` and `failWith`

* Remove now obsolete `epsilonMaybeEq` from `Utils`

Lucsanszky committed:
Use `boundRational` and `unboundRational` methods

Avoid using the functions that are used to implement them directly

Lucsanszky commented on libs/cardano-ledger-core/test/Test/Cardano/Ledger/BaseTypesSpec.hs:
```diff
@@ -56,29 +55,16 @@ boundedRationalTests badJSONValues = do
               Left err -> error err
-              Right (br' :: a) ->
-                epsilonMaybeEq 1e-18 (unboundRational br) (unboundRational br')
-      prop "Roundtrip to Scientific and back up to an epsilon" $
```
I removed these tests since I believe my changes make them obsolete.

Lucsanszky commented on libs/cardano-ledger-core/test/Test/Cardano/Ledger/BaseTypesSpec.hs:
```diff
@@ -17,7 +17,6 @@ import Data.Typeable
 import Test.Cardano.Ledger.Binary.RoundTrip (roundTripCborExpectation)
 import Test.Cardano.Ledger.Common
 import Test.Cardano.Ledger.Core.Arbitrary ()
-import Test.Cardano.Ledger.Core.Utils (epsilonMaybeEq)
```
`epsilonMaybeEq` was only used here, so we might want to remove the function itself as well. Unless we think that it might be useful again at some point.

lehins replied:
I think we can remove it. If we ever need it again, we can always bring it back. I have a variant of it in a couple of projects of mine.
I doubt though that we will need it, since it is usually only useful for testing floating point numbers and other computation related to rounding, which is unlikely to be useful in ledger again.

Lucsanszky commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -328,53 +335,42 @@ instance
       fromRationalEither (coeff % (10 ^ negate exp10))
   | otherwise = do
-      when (19 < exp10) $ failWith "too big"
+      when (19 < exp10) $ failWith "too big" sci
```
`19` is the number of decimal digits for 64 bit values. I think it might be a good idea to consider removing this magic number by introducing a constant. Something along the lines of `numberOfDecimals64 = 19`. Would that be desirable? If so, I can do it here and now or later as part of some refactoring.

lehins replied:
I totally agree with you. Let's create a top level binding with a bit of explanation what this number means. I think doing it in this PR will be good.

Lucsanszky reviewed (commented):
This is now ready for review. Added some comments to consider.

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -244,6 +245,12 @@ instance Integral a => Ord (BoundedRatio b a) where
 
+toRationalJSON :: Rational -> Value
+toRationalJSON r =
+  case fromRationalRepetendLimited 20 r of
```
I believe we need to use `19` here as well, otherwise there could be issues with roundtripping
```suggestion
  case fromRationalRepetendLimited 19 r of
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -328,53 +335,42 @@ instance
+fromRationalEither r = maybe (failWith "outside of bounds" r) Right $ boundRational r
+
+failWith :: Show a => String -> a -> Either String b
+failWith msg (show -> val) = Left $ "Value is " <> msg <> ": " <> val
```
I'd suggest using `ViewPatterns` only in places where they are really useful. Some people get triggered about their usage. (none on our team, from what I know, but it does happen in Haskell community)
```suggestion
failWith msg val = Left $ "Value is " <> msg <> ": " <> show val
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ExUnits.hs:
```diff
@@ -32,6 +32,7 @@ import Cardano.Ledger.BaseTypes (
   BoundedRational (unboundRational),
   NonNegativeInterval,
   boundRational,
+  toRationalJSON,
```
We no longer need to import it here, because this comment is no longer true:
https://github.com/IntersectMBO/cardano-ledger/blob/21871e9651d8d3fad3383647b4f157f1612ae685/libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ExUnits.hs#L172

So, we can fix the ToJSON/FromJSON instances for `Prices`, by relying on the underlying instance of `NonNegativeInterval`
```suggestion
```

lehins reviewed (changes requested):
Very nice!

I left one suggestion that I didn't even think of until reviewing this PR. We can now simplify JSON instances for `Prices`.

Once that's done we can merge the PR

Lucsanszky committed:
Fix `ToJSON`/`FromJSON` instances for `Prices`

We can rely on the JSON instances of `NonNegativeInterval`

Lucsanszky committed:
Introduce `maxDecimalsWord64`

The maximum precision that we can achieve with `Word64` is `20` digits,
hence the maximum scale / decimals,
the number of digits after the decimal point, is `19`.
We often use this number during `Scientific` conversions and thus
it should be represented as a top-level binding for clarity.

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -176,6 +177,9 @@ import NoThunks.Class (NoThunks (..))
 import Quiet (Quiet (Quiet))
 
+maxScale :: Int
+maxScale = 19
```
We should not export this biding from the module and it should have a better name, along the line of:
(Edit: tests can define their own binding like that)
```suggestion
numFullDigitsWord64 :: Int
numFullDigitsWord64 = 19
```

Lucsanszky replied:
Renamed it. The "full digits" part sounds strange to me but including `Word64` in the name is good. `maxScale` could indeed be a bit misleading without context.

lehins replied:
`maxBound` for `Word64` has 20 digits:
```haskell
ghci> length (show (maxBound :: Word))
20
```

But with regards to representing decimal numbers it has only 19 digits available:
```
ghci> (2 * 10 ^ 19) > toInteger (maxBound :: Word)
True
ghci> (1 * 10 ^ 19) > toInteger (maxBound :: Word)
False
```
For example, the smallest decimal number you can represent with `Ratio Word64` is this number that has 19 decimal places (despite that numerator has 20 digits :grin:):
```
ghci> 0.0000000000000000001 :: Ratio Word
1 % 10000000000000000000
ghci> 0.00000000000000000001 :: Ratio Word
1 % 7766279631452241920
```

With all this information at your disposal, if you have a better name for the binding, please go for it.

Lucsanszky replied:
Yes, this is what I added to the commit body here: be55cbc14d2fc61970b711d77eb857b70e8c896b . Although not in such detail. :grin:  This is why I picked `maxScale` since you refer to that parameter as scale too in `safe-decimal`: https://hackage.haskell.org/package/safe-decimal-0.2.1.0/docs/Numeric-Decimal.html

How about `maxScaleWord64`? Or `maxDecimalsWord64` since mathematics sometimes refer to the decimal places after the decimal point as decimals? I think this latter one is more well-known.

lehins replied:
`maxDecimalsWord64` sounds ideal!

Lucsanszky committed:
Remove unused extensions

lehins reviewed (approved):
Perfect!
